### ISSUE TITLE
Incomplete installation of optional dependencies

### ISSUE THREAD

**Issue by hoefling** [sdispater/poetry#1121]
- [x] I am on the [latest](https://github.com/sdispater/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/sdispater/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: Gentoo Linux
- **Poetry version**: 0.12.16

### `pyproject.toml`
<details><summary>Click</summary>

```toml
[tool.poetry]
name = "spam"
version = "0.1.0"
description = ""
authors = ["Oleg Höfling <oleg.hoefling@gmail.com>"]

[tool.poetry.dependencies]
python = '~2.7 || ^3.4'
keyring = { version = "*", optional = true }

[tool.poetry.extras]
keyring = ['keyring']

[tool.poetry.dev-dependencies]

[build-system]
requires = ["poetry>=0.12"]
build-backend = "poetry.masonry.api"
```
</details>

## Issue

`poetry install` doesn't install the complete list of dependencies. The optional package `keyring`
 requires Freedesktop's `secretstorage`, the latest version of which (>3.0) requires `python>=3.5` and `jeepney`. However, running `poetry install -vvv --extras keyring` yields: <details><summary>Click</summary>

```
$ python -V
Python 3.6.8
$ pip -V
pip 19.1.1 from /home/hoefling/projects/.venvs/test/lib/python3.6/site-packages/pip (python 3.6)
$ poetry -V
Poetry 0.12.16
$ poetry install -vvv --extras keyring
Updating dependencies
Resolving dependencies...
   1: fact: spam is 0.1.0
   1: derived: spam
   1: fact: spam depends on keyring (*)
   1: selecting spam (0.1.0)
   1: derived: keyring (*)
PyPI: 137 packages found for keyring *
   1: fact: keyring (19.0.2) requires Python >=3.5
   1: derived: not keyring (19.0.2)
   1: fact: keyring (19.0.1) requires Python >=3.5
   1: derived: not keyring (19.0.1)
   1: fact: keyring (19.0.0) requires Python >=3.5
   1: derived: not keyring (19.0.0)
   0: Duplicate dependencies for secretstorage
   0: Different requirements found for secretstorage (<3) and secretstorage (*).
   1: Version solving took 0.287 seconds.
   1: Tried 1 solutions.
   0: Retrying dependency resolution for Python (>=2.7,<2.8 || >=3.4,<3.5).
   1: fact: spam is 0.1.0
   1: derived: spam
   1: fact: spam depends on keyring (*)
   1: selecting spam (0.1.0)
   1: derived: keyring (*)
   1: fact: keyring (19.0.2) requires Python >=3.5
   1: derived: not keyring (19.0.2)
   1: fact: keyring (19.0.1) requires Python >=3.5
   1: derived: not keyring (19.0.1)
   1: fact: keyring (19.0.0) requires Python >=3.5
   1: derived: not keyring (19.0.0)
   1: fact: keyring (18.0.1) depends on entrypoints (*)
   1: fact: keyring (18.0.1) depends on secretstorage (<3)
   1: fact: keyring (18.0.1) depends on pywin32-ctypes (<0.1.0 || >0.1.0,<0.1.1 || >0.1.1)
   1: selecting keyring (18.0.1)
   1: derived: pywin32-ctypes (<0.1.0 || >0.1.0,<0.1.1 || >0.1.1)
   1: derived: secretstorage (<3)
   1: derived: entrypoints (*)
PyPI: 3 packages found for pywin32-ctypes <0.1.0 || >0.1.0,<0.1.1 || >0.1.1
PyPI: 12 packages found for secretstorage <3
PyPI: 6 packages found for entrypoints *
   1: selecting pywin32-ctypes (0.2.0)
   1: fact: entrypoints (0.3) depends on configparser (>=3.5)
   1: selecting entrypoints (0.3)
   1: derived: configparser (>=3.5)
PyPI: 9 packages found for configparser >=3.5
   1: selecting configparser (3.7.4)
   1: fact: secretstorage (2.3.1) depends on cryptography (*)
   1: selecting secretstorage (2.3.1)
   1: derived: cryptography (*)
PyPI: 71 packages found for cryptography *
   1: fact: cryptography (2.6.1) depends on asn1crypto (>=0.21.0)
   1: fact: cryptography (2.6.1) depends on six (>=1.4.1)
   1: fact: cryptography (2.6.1) depends on cffi (>=1.8,<1.11.3 || >1.11.3)
   1: fact: cryptography (2.6.1) depends on enum34 (*)
   1: fact: cryptography (2.6.1) depends on ipaddress (*)
   1: selecting cryptography (2.6.1)
   1: derived: ipaddress (*)
   1: derived: enum34 (*)
   1: derived: cffi (>=1.8,<1.11.3 || >1.11.3)
   1: derived: six (>=1.4.1)
   1: derived: asn1crypto (>=0.21.0)
PyPI: 21 packages found for ipaddress *
PyPI: 27 packages found for enum34 *
PyPI: 14 packages found for cffi >=1.8,<1.11.3 || >1.11.3
PyPI: 15 packages found for six >=1.4.1
PyPI: 5 packages found for asn1crypto >=0.21.0
   1: selecting asn1crypto (0.24.0)
   1: fact: cffi (1.12.3) depends on pycparser (*)
   1: selecting cffi (1.12.3)
   1: derived: pycparser (*)
PyPI: 19 packages found for pycparser *
   1: selecting six (1.12.0)
   1: selecting pycparser (2.19)
   1: selecting ipaddress (1.0.22)
   1: selecting enum34 (1.1.6)
   1: Version solving took 0.180 seconds.
   1: Tried 1 solutions.
   0: Retrying dependency resolution for Python (>=3.5,<4.0).
   1: fact: spam is 0.1.0
   1: derived: spam
   1: fact: spam depends on keyring (*)
   1: selecting spam (0.1.0)
   1: derived: keyring (*)
   1: fact: keyring (19.0.2) depends on entrypoints (*)
   1: fact: keyring (19.0.2) depends on secretstorage (*)
   1: fact: keyring (19.0.2) depends on pywin32-ctypes (<0.1.0 || >0.1.0,<0.1.1 || >0.1.1)
   1: selecting keyring (19.0.2)
   1: derived: pywin32-ctypes (<0.1.0 || >0.1.0,<0.1.1 || >0.1.1)
   1: derived: secretstorage (*)
   1: derived: entrypoints (*)
PyPI: 15 packages found for secretstorage *
   1: selecting pywin32-ctypes (0.2.0)
   1: selecting entrypoints (0.3)
   1: fact: secretstorage (3.1.1) depends on cryptography (*)
   1: fact: secretstorage (3.1.1) depends on jeepney (*)
   1: selecting secretstorage (3.1.1)
   1: derived: jeepney (*)
   1: derived: cryptography (*)
PyPI: 5 packages found for jeepney *
   1: selecting jeepney (0.4)
   1: fact: cryptography (2.6.1) depends on asn1crypto (>=0.21.0)
   1: fact: cryptography (2.6.1) depends on six (>=1.4.1)
   1: fact: cryptography (2.6.1) depends on cffi (>=1.8,<1.11.3 || >1.11.3)
   1: selecting cryptography (2.6.1)
   1: derived: cffi (>=1.8,<1.11.3 || >1.11.3)
   1: derived: six (>=1.4.1)
   1: derived: asn1crypto (>=0.21.0)
   1: selecting asn1crypto (0.24.0)
   1: fact: cffi (1.12.3) depends on pycparser (*)
   1: selecting cffi (1.12.3)
   1: derived: pycparser (*)
   1: selecting six (1.12.0)
   1: selecting pycparser (2.19)
   1: Version solving took 0.091 seconds.
   1: Tried 1 solutions.
   0: Complete version solving took 0.567 seconds for 2 branches
   0: Resolved for branches: (>=2.7,<2.8 || >=3.4,<3.5), (>=3.5,<4.0)

Writing lock file


Package operations: 7 installs, 0 updates, 0 removals, 3 skipped

  - Installing pycparser (2.19)
  - Installing asn1crypto (0.24.0)
  - Installing cffi (1.12.3)
  - Skipping six (1.12.0) Already installed
  - Installing cryptography (2.6.1)
  - Skipping jeepney (0.4) Not required
  - Installing entrypoints (0.3)
  - Skipping pywin32-ctypes (0.2.0) Not needed for the current environment
  - Installing secretstorage (3.1.1)
  - Installing keyring (19.0.2)
```
</details>

`jeepney` shouldn't be skipped as all the requirement conditions are met. In fact, I suppose this has smth to do with the `optional` keyword - if I remove `keyring` extras:

```toml
[tool.poetry.dependencies]
python = '~2.7 || ^3.4'
keyring = '*'
```
then all deps including `jeepney` are installed just fine.
